server/datastore/s3/carves.go

```diff
@@ -0,0 +1,236 @@
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/fleetdm/fleet/server/kolide"
+	"github.com/pkg/errors"
+)
+
+const (
+	defaultMaxS3Keys = 1000
+	cleanupSize      = 1000
+	// This is Golang's way of formatting timestrings, it's confusing, I know.
+	// If you are used to more conventional timestrings, this is equivalent
+	// to %Y/%m/%d/%H (year/month/day/hour)
+	timePrefixFormat = "2006/01/02/15"
+)
+
+// generateS3Key builds S3 key from carve metadata
+// all keys are prefixed by date so that they can easily be listed chronologically
+func (d *Datastore) generateS3Key(metadata *kolide.CarveMetadata) string {
+	simpleDateHour := metadata.CreatedAt.Format(timePrefixFormat)
+	return fmt.Sprintf("%s%s/%s", d.prefix, simpleDateHour, metadata.Name)
+}
+
+// NewCarve initializes a new file carving session
+func (d *Datastore) NewCarve(metadata *kolide.CarveMetadata) (*kolide.CarveMetadata, error) {
+	objectKey := d.generateS3Key(metadata)
+	res, err := d.s3client.CreateMultipartUpload(&s3.CreateMultipartUploadInput{
+		Bucket: &d.bucket,
+		Key:    &objectKey,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "s3 multipart carve create")
+	}
+	metadata.SessionId = *res.UploadId
+	return d.metadatadb.NewCarve(metadata)
+}
+
+// UpdateCarve updates carve definition in database
+// Only max_block and expired are updatable
+func (d *Datastore) UpdateCarve(metadata *kolide.CarveMetadata) error {
+	return d.metadatadb.UpdateCarve(metadata)
+}
+
+// listS3Carves lists all keys up to a given one or if the passed max number
+// of keys has been reached; keys are returned in a set-like map
+func (d *Datastore) listS3Carves(lastPrefix string, maxKeys int) (map[string]bool, error) {
+	var err error
+	var continuationToken string
+	result := make(map[string]bool)
+	if maxKeys <= 0 {
+		maxKeys = defaultMaxS3Keys
+	}
+	if !strings.HasPrefix(lastPrefix, d.prefix) {
+		lastPrefix = d.prefix + lastPrefix
+	}
+	for {
+		carveFilesPage, err := d.s3client.ListObjectsV2(&s3.ListObjectsV2Input{
+			Bucket:            &d.bucket,
+			Prefix:            &d.prefix,
+			ContinuationToken: &continuationToken,
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, carveObject := range carveFilesPage.Contents {
+			result[*carveObject.Key] = true
+			if strings.HasPrefix(*carveObject.Key, lastPrefix) || len(result) >= maxKeys {
+				return result, nil
+			}
+		}
+		if !*carveFilesPage.IsTruncated {
+			break
+		}
+		continuationToken = *carveFilesPage.ContinuationToken
+	}
+	return result, err
+}
+
+// CleanupCarves is a noop on the S3 side since users should rely on the bucket
+// lifecycle configurations provided by AWS. This will compare a portion of the
+// metadata present in the database and mark as expired the carves no longer
+// available in S3 (ignores the `now` argument)
+func (d *Datastore) CleanupCarves(now time.Time) (int, error) {
+	var err error
+	// Get the 1000 oldest carves
```

Comment:
After 1000 carves will this no longer track the status of the newer carves?

Reply:
The idea is that 99% of the time the bucket life-cycle configuration will be setup to simply delete objects older than X days.
If that holds true, since the cleanup runs in the background every hour, it can just check if the 1000 oldest and non-expired carves in the metadata DB are still present in S3, and if not mark them as expired, eventually converging with what is actually present in S3. Of course, this is far from perfect, for example if more than 24k carve are collected (and then deleted from S3) in a single day, the convergence will take longer than a day, and if that same load is repeat every day, it will never converge. I was planning to add these considerations as part of the documentation instead of trying to over-complicate the syncing mechanism.

Reply:
Yeah this makes sense to me now that I've read the documentation. Thank you.